python/TestHarness/testers/bench.py

```diff
@@ -0,0 +1,330 @@
+
+        self.dur_secs = end - start
+
+        # write perflog
+        if self.haveperflog:
+            with open(os.path.join(tmpdir, 'perflog.csv'), 'r') as csvfile:
+                reader = csv.reader(csvfile)
+                skip = True # use to skip header line
+                for row in reader:
+                    if not skip:
+                        self.perflog.append(row)
+                    else:
+                        skip = False
+
+        shutil.rmtree(tmpdir)
+
+class Benchmark(Tester):
+    @staticmethod
+    def validParams():
+        params = Tester.validParams()
+        params.addParam('input',              'The input file to use for this test.')
+        params.addParam('test_name',          'The name of the test - populated automatically')
+        params.addParam('cumulative_dur', 60, 'cumulative time (secs) to run each benchmark')
+        params.addParam('min_runs', 10,       'minimum number of runs for each benchmark')
+        return params
+
+    def __init__(self, name, params):
+        Tester.__init__(self, name, params)
+        self.tags.append('benchmarks')
+
+        self.params = params
+        self.benchmark = None
+        self.db = 'benchmarks.sqlite'
+
+    # override
+    def checkRunnable(self, options):
+        if 'benchmarks' not in options.runtags:
+            self.setStatus('benchmark', self.bucket_skip)
+            return False
+        return True
+
+    # override
+    def run(self, timer, options):
+        p = self.params
+        if options.method not in ['opt', 'oprof']:
+            raise ValueError('cannot run benchmark with "' + options.method + '" build')
+        t = Test(p['executable'], p['input'], args=p['cli_args'], rootdir=p['test_dir'])
+        self.benchmark = Bench(p['test_name'], test=t, cum_dur=p['cumulative_dur'], min_runs=p['min_runs'])
+
+        self.benchmark.run(timer)
+
+        with DB(self.db) as db:
+            self.benchmark.run()
+            db.store(self.benchmark)
+
+    # override
+    def processResults(self, moose_dir, options, output):
+        self.setStatus(self.exit_code, self.bucket_success)
+        return output
+
+class Bench:
+    def __init__(self, name, realruns=None, test=None, cum_dur=60, min_runs=10, max_runs=1000):
+        self.name = name
+        self.test = test
+        self.realruns = []
+        self.perflogruns = []
+        if realruns is not None:
+            self.realruns.extend(realruns)
+        self._cum_dur = cum_dur
+        self._min_runs = min_runs
+        self._max_runs = max_runs
+
+    def run(self, timer=None):
+        # TODO: also collect MOOSE perflog data
+        tot = 0.0
+        while True:
+            self.test.run(timer)
+            self.realruns.append(self.test.dur_secs)
+            self.perflogruns.append(self.test.perflog)
+            tot += self.test.dur_secs
+            if (len(self.realruns) >= self._min_runs and tot >= self._cum_dur) or len(self.realruns) >= self._max_runs:
```

Comment:
You know you could just put this condition in the while loop. I'm not sure why you have a separate "if" statement with a break at the bottom of the loop. This doesn't appear to be a do/while loop to me.